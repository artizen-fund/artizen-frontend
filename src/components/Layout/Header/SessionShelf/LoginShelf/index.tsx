import { useState } from 'react'
import { useApolloClient, ApolloClient } from '@apollo/client'
import styled from 'styled-components'
import { Icon, Form, CheckboxControl, Button } from '@components'
import { loginWithEmail, useMagic, useFormLocalStorage } from '@lib'
import { breakpoint } from '@theme'
import { schema, uischema, initialState, FormState } from '@forms/login'
import {
  Copy,
  Headline,
  SignInDirections,
  InfoRow,
  Confirmation,
  Reset,
  CheckWrapper,
  Check,
  CheckMessage,
  ISessionShelf,
} from '../_common'

const LoginShelf = ({ setCreateMode }: ISessionShelf) => {
  const LOCALSTORAGE_KEY = 'loginForm'
  const [data, setData] = useFormLocalStorage<FormState>(LOCALSTORAGE_KEY, initialState)

  const { magic } = useMagic()
  const apolloClient = useApolloClient()

  const [sentEmail, setSentEmail] = useState(false)
  const [readonly, setReadonly] = useState(false)
  const [recordEmail, setRecordEmail] = useState(true)

  const handleEmailLogin = async (apolloClient: ApolloClient<object>, email?: string, magic?: MagicInstance) => {
    if (!magic) {
      throw 'Error: magic session not initialized.'
    }
    if (!email) {
      throw 'Error: email is missing'
    }
    setReadonly(true)
    try {
      setSentEmail(true)
      await loginWithEmail(apolloClient, magic, email)
      setReadonly(false)
    } catch (error) {
      console.error(error)
      setReadonly(false)
    }
  }

  const reset = () => {
    setReadonly(false)
  }

  return (
    <Wrapper className={readonly ? 'submitted' : ''}>
      <Copy>
        <Headline>Sign in using the email associated with your Artizen account</Headline>
        <InfoRow onClick={() => setCreateMode(true)}>
          <Icon glyph="infoLarge" outline level={1} />
          <SignInDirections>Sign up for an account instead</SignInDirections>
        </InfoRow>
      </Copy>
      <Form localStorageKey={LOCALSTORAGE_KEY} {...{ schema, uischema, initialState, data, setData, readonly }}>
        <>
          <Buttons>
            <Button stretch onClick={() => handleEmailLogin(apolloClient, data.email, magic)}>
              Sign In
            </Button>
            <Button transparent stretch onClick={() => setCreateMode(true)}>
              Sign Up instead
            </Button>
          </Buttons>

          {sentEmail && (
            <Confirmation>
              <Icon glyph="tick" outline level={2} color="moss" />
              <div>
                <h1>Done, confirmation sent!</h1>
                <p>
                  We emailed a magic link to {data.email}.<br />
                  Click the link sign in.
                </p>
              </div>
              <Reset onClick={() => reset()}>Didnâ€™t receive an email?</Reset>
            </Confirmation>
          )}
        </>
      </Form>
      <CheckWrapper>
        <Check>
          <CheckboxControl
            data={recordEmail}
            path="not-used"
            handleChange={() => setRecordEmail(!recordEmail)}
            label=""
          />
          <CheckMessage>Remember my email address.</CheckMessage>
        </Check>
      </CheckWrapper>
    </Wrapper>
  )
}

const Buttons = styled.div`
  grid-area: buttons;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
`

const Wrapper = styled.div`
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-areas:
      'copy email'
      'copy buttons'
      'tocCheck buttons';
    &.submitted {
      grid-template-areas:
        'copy confirmation'
        'copy confirmation'
        'tocCheck confirmation';
    }
    gap: 10px;

    .vertical-layout,
    .vertical-layout-item {
      display: contents;
    }

    *[id='#/properties/email'] {
      grid-area: email;
    }
  }

  &.submitted {
    *[id='#/properties/email'],
    ${Buttons} {
      display: none;
    }
    ${Confirmation} {
      display: flex;
    }
  }
`

export default LoginShelf
