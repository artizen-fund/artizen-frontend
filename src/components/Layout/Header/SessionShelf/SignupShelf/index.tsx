import { useEffect, useState, useContext } from 'react'
import { useApolloClient, ApolloClient } from '@apollo/client'
import Link from 'next/link'
import styled from 'styled-components'
import { Icon, Form, CheckboxControl } from '@components'
import { loginWithEmail, useMagic, useFormLocalStorage, UserContext } from '@lib'
import { UPDATE_NEW_USER_PROFILE } from '@gql'
import { breakpoint } from '@theme'
import { schema, uischema, initialState, FormState } from '@forms/signup'
import {
  Copy,
  Headline,
  SignInDirections,
  InfoRow,
  SubmitButton,
  Confirmation,
  Reset,
  CheckWrapper,
  Check,
  CheckMessage,
  ISessionShelf,
} from '../_common'
import { IUser } from '@types'

const SignupShelf = ({ setCreateMode }: ISessionShelf) => {
  const LOCALSTORAGE_KEY = 'signupForm'
  const [data, setData] = useFormLocalStorage<FormState>(LOCALSTORAGE_KEY, initialState)

  const { magic } = useMagic()
  const apolloClient = useApolloClient()
  const { loggedInUser } = useContext(UserContext)

  const [sentEmail, setSentEmail] = useState(false)
  const [submitted, setSubmitted] = useState(false)
  const [readonly, setReadonly] = useState(false)
  const [acceptedToc, setAcceptedToc] = useState(true)

  const handleEmailLogin = async (apolloClient: ApolloClient<object>, email?: string, magic?: MagicInstance) => {
    if (!magic) {
      throw 'Error: magic session not initialized.'
    }
    if (!email) {
      throw 'Error: email is missing'
    }
    setReadonly(true)
    setSubmitted(true)
    try {
      setSentEmail(true)
      await loginWithEmail(apolloClient, magic, email)
      setReadonly(false)
    } catch (error) {
      console.error(error)
      setReadonly(false)
    }
  }

  const reset = () => {
    setSubmitted(false)
    setReadonly(false)
  }

  const saveNewUserData = async (user: IUser, data: FormState) => {
    try {
      await apolloClient.mutate({
        mutation: UPDATE_NEW_USER_PROFILE,
        variables: { id: user.id, firstName: data.firstName, lastName: data.lastName },
      })
    } catch (error) {
      console.error('Error saving new user profile', error)
    }
  }

  useEffect(() => {
    return () => {
      if (!!loggedInUser && !!data.firstName && !!data.lastName) {
        saveNewUserData(loggedInUser, data)
      }
    }
  }, [data, loggedInUser])

  return (
    <Wrapper className={submitted ? 'submitted' : ''}>
      <Copy>
        <Headline>Let’s get started by setting up your Artizen account </Headline>
        <InfoRow onClick={() => setCreateMode(false)}>
          <Icon glyph="infoLarge" outline level={1} />
          <SignInDirections>Sign in to your account instead</SignInDirections>
        </InfoRow>
      </Copy>
      <Form
        submitDisabledFromOutside={!acceptedToc || submitted}
        localStorageKey={LOCALSTORAGE_KEY}
        {...{ schema, uischema, initialState, data, setData, readonly }}
      >
        <>
          <SubmitButton stretch onClick={() => handleEmailLogin(apolloClient, data.email, magic)}>
            Sign In / Sign Up
          </SubmitButton>
          {sentEmail && (
            <Confirmation>
              <Icon glyph="tick" outline level={2} color="moss" />
              <div>
                <h1>Done, confirmation sent!</h1>
                <p>
                  We emailed a magic link to {data.email}.<br />
                  Click the link Sign in or sign up.
                </p>
              </div>
              <Reset onClick={() => reset()}>Didn’t receive an email?</Reset>
            </Confirmation>
          )}
        </>
      </Form>
      <CheckWrapper>
        <Check>
          <CheckboxControl
            data={acceptedToc}
            path="not-used"
            handleChange={() => setAcceptedToc(!acceptedToc)}
            label=""
          />
          <CheckMessage>
            I agree to Artizen’s{' '}
            <Link href="/toc">
              <a>Terms &amp; Conditions</a>
            </Link>
          </CheckMessage>
        </Check>
      </CheckWrapper>
    </Wrapper>
  )
}

const Wrapper = styled.div`
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    display: grid;
    grid-template-columns: 2fr 1fr 1fr;
    grid-template-areas:
      'copy firstName lastName'
      'copy email email'
      'tocCheck submit submit';
    &.submitted {
      grid-template-areas:
        'copy confirmation confirmation'
        'copy confirmation confirmation'
        'tocCheck confirmation confirmation';
    }
    gap: 10px;

    .vertical-layout,
    .vertical-layout-item {
      display: contents;
    }

    *[id='#/properties/email'] {
      grid-area: email;
    }
    *[id='#/properties/firstName'] {
      grid-area: firstName;
    }
    *[id='#/properties/lastName'] {
      grid-area: lastName;
    }
  }

  &.submitted {
    *[id='#/properties/firstName'],
    *[id='#/properties/lastName'],
    *[id='#/properties/email'],
    ${SubmitButton} {
      display: none;
    }
    ${Confirmation} {
      display: flex;
    }
  }
`

export default SignupShelf
