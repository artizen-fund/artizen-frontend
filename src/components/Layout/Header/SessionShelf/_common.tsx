import styled from 'styled-components'
import { rgba } from '@lib'
import { palette, typography, breakpoint } from '@theme'

export const Copy = styled.div`
  grid-area: copy;
  display: flex;
  flex-direction: column;
  gap: 15px;
`

export const Headline = styled.h1`
  ${typography.title.l2};
`

export const InfoRow = styled.div`
  display: flex;
  flex-direction: row;
  gap: 10px;
  justify-content: flex-start;
  align-items: center;
  margin: 1em 0 2em 0;
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    margin: 1em 0 0 0;
  }
`

export const SignInDirections = styled.p`
  ${typography.label.l1};
`

export const CheckWrapper = styled.div`
  display: contents;
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    grid-area: tocCheck;
  }
`

export const Check = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: ${breakpoint.laptop}px) {
    justify-content: start;
  }
  a {
    text-decoration: underline;
    text-decoration-thickness: 2px;
  }
`

export const CheckMessage = styled.p`
  ${typography.label.l1}
  color: ${rgba(palette.night)};
`

export const Confirmation = styled.div`
  display: none;
  grid-area: confirmation;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 30px;
  h1 {
    ${typography.title.l4}
    color: ${rgba(palette.night)};
    @media (prefers-color-scheme: dark) {
      color: ${rgba(palette.moon)};
    }
  }
  p {
    ${typography.label.l1}
    color: ${rgba(palette.barracuda)};
  }
  text-align: center;
`

export const Reset = styled.p`
  cursor: pointer;
  border-bottom: 2px solid ${rgba(palette.night)};
  @media (prefers-color-scheme: dark) {
    border-bottom: 2px solid ${rgba(palette.moon)};
  }
`
