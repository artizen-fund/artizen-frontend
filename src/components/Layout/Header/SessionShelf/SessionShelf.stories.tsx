import SessionShelf from './'

export default {
  title: 'header/SessionShelf',
  component: SessionShelf,
  argTypes: {},
}

export const SessionShelfComponent = (props: any) => <SessionShelf {...props} />
