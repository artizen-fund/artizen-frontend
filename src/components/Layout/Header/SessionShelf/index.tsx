import { useSession } from '@lib'
import LoginShelf from '../LoginShelf'
import AccountShelf from '../AccountShelf'

const SessionShelf = () => {
  const { user } = useSession()

  return !!user ? <AccountShelf {...{ user }} /> : <LoginShelf />
}

export default SessionShelf
