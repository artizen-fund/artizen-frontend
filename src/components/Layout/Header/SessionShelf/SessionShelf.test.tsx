import renderer from 'react-test-renderer'
import { SessionProvider } from '@lib'
import SessionShelf from './'

describe('SessionShelf', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <SessionProvider>
          <SessionShelf />
        </SessionProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
