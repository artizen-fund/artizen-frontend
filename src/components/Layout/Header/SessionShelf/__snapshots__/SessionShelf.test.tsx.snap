// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionShelf renders correctly 1`] = `
<div
  className="sc-4a4f802d-3 fuQvvd"
>
  <div
    className="sc-4a4f802d-4 fhjysv"
  >
    <h1
      className="sc-4a4f802d-5 jbsSsK"
    >
      Let’s get started by setting up your Artizen account
    </h1>
    <div
      className="sc-4a4f802d-6 ehlijM"
    >
      <div
        className="sc-8c7bf1ae-2 huxvCL"
      >
        <div
          className="sc-8c7bf1ae-0 bhoRgU"
          color="night"
        >
          <div
            className="sc-1d011157-0 ejjWfH"
            color="night"
          />
        </div>
      </div>
      <p
        className="sc-4a4f802d-7 hAzDhs"
      >
        Already have an account?
        <br />
        You can still use this form, it’s magic!
      </p>
    </div>
  </div>
  <div
    className="vertical-layout"
    hidden={false}
  >
    <div
      className="vertical-layout-item"
    >
      <div
        className="sc-2f9d5146-0 jjUILZ"
        id="#/properties/email"
      >
        <div
          className="sc-dccedff6-0 dxDPQx"
        >
          <input
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="e.g. rene@artizen.fund"
            required={true}
            type="text"
          />
          <label
            className="sc-ffdb2e6c-0 dXVxRc"
          >
            Email Address
             *
          </label>
        </div>
        <div
          className="sc-93ba83c0-0 bAVzHI"
        />
      </div>
    </div>
  </div>
  <button
    className="sc-a6b3add2-2 ebsYdi sc-4a4f802d-0 iErgSy  disabled"
    disabled={true}
    onClick={[Function]}
  >
    <span>
      Sign In / Sign Up
    </span>
  </button>
  <div
    className="sc-4a4f802d-2 hPhXmc"
  >
    <div>
      Congrats, confirmation sent!
    </div>
    <p>
      We emailed a magic link to 
      .
    </p>
  </div>
  <div
    className="sc-4a4f802d-1 iqsVwz"
  >
    <div
      className="sc-4a4f802d-8 hyNfNI"
    />
    <div
      className="sc-4a4f802d-9 iCNaSc"
    >
      <button
        className="sc-a6b3add2-2 erODIt undefined "
        onClick={[Function]}
      >
        <span>
          Phone
        </span>
      </button>
      <button
        className="sc-a6b3add2-2 erODIt undefined "
        onClick={[Function]}
      >
        <span>
          Twitter
        </span>
      </button>
      <button
        className="sc-a6b3add2-2 erODIt undefined "
        onClick={[Function]}
      >
        <span>
          Discord
        </span>
      </button>
    </div>
  </div>
  <div
    className="sc-4a4f802d-10 kwsLrN"
  >
    <div
      className="sc-4a4f802d-11 fRBTKZ"
    >
      <label
        className="sc-dd35d48e-4 dCKwdz"
      >
        <div
          className="sc-dd35d48e-0 lmHjtw"
        >
          <input
            checked={true}
            className="sc-dd35d48e-1 clTMzK"
            disabled={false}
            onChange={[Function]}
            required={false}
            type="checkbox"
          />
          <span
            className="sc-dd35d48e-2 dtceeR"
          />
        </div>
        <span
          className="sc-dd35d48e-3 eAaGSv"
        />
      </label>
      <p
        className="sc-4a4f802d-12 camNOd"
      >
        I agree to Artizen’s
        <a
          href="/toc"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
           Terms & Conditions
        </a>
      </p>
    </div>
  </div>
</div>
`;
